backend/editor/controllers/search_controller.py

```diff
@@ -0,0 +1,374 @@
+
+    if search_query is None:
+        return None
+
+    if language_code is not None:
+        search_query = f"tags_ids_{language_code}:{search_query}"
+
+    return search_query
+
+
+def build_cypher_name_search_term(project_id: str) -> tuple[str, str, str]:
+    query = f"""
+            CALL db.index.fulltext.queryNodes("{project_id}_SearchTagsIds", $search_query)
+            YIELD node, score
+            WHERE score > 0.1
+            WITH node.id AS nodeId
+            WITH COLLECT(nodeId) AS nodeIds
+        """
+
+    where_clause = "n.id IN nodeIds"
+
+    order_clause = "WITH n, apoc.coll.indexOf(nodeIds, n.id) AS index ORDER BY index"
+
+    return query, where_clause, order_clause
+
+
+def build_cypher_query(
+    query: Query, skip: int, limit: int
+) -> tuple[str, str, dict[str, str]] | None:
+    lucene_name_search_queries = list(
+        filter(
+            lambda q: q is not None, map(build_lucene_name_search_query, query.name_search_terms)
+        )
+    )
+
+    cypher_filter_search_terms = [
+        term.build_cypher_query(index) for index, term in enumerate(query.filter_search_terms)
+    ]
+
+    full_text_search_query, order_clause = "", ""
+    query_params = {}
+
+    if lucene_name_search_queries:
+        (
+            full_text_search_query,
+            name_filter_search_term,
+            order_clause,
+        ) = build_cypher_name_search_term(query.project_id)
+        query_params["search_query"] = " AND ".join(lucene_name_search_queries)
+        cypher_filter_search_terms.append((name_filter_search_term, None))
+
+    query_params |= {
+        get_query_param_name(index): cypher_filter_search_term[1]
+        for index, cypher_filter_search_term in enumerate(cypher_filter_search_terms)
+    }
+
+    combined_filter_query = (
+        f"WHERE {' AND '.join(map(lambda x: x[0], cypher_filter_search_terms))}"
+        if cypher_filter_search_terms
+        else ""
+    )
+
+    base_query = f"""
+    {full_text_search_query}
+    MATCH (n:{query.project_id}:ENTRY)
+    {combined_filter_query}
+    """
+
+    page_subquery = f"""
+    {order_clause}
+    WITH collect(n) AS nodeList, count(n) AS nodeCount
+    UNWIND nodeList AS node
+    WITH node, nodeCount
+    SKIP {skip} LIMIT {limit}
+    WITH collect(node) AS nodeList, nodeCount
+    RETURN nodeList, nodeCount;
+    """
+
+    count_subquery = """
+    RETURN count(n) AS nodeCount;
+    """
+
+    page_query = base_query + page_subquery
+    count_query = base_query + count_subquery
+
+    return page_query, count_query, query_params
+
+
+async def search_entry_nodes(project_id: str, raw_query: str, page: int) -> EntryNodeSearchResult:
+    """
+    Search for entry nodes in the database
+    """
+    query = validate_query(project_id, raw_query)
+
+    if query is None:
+        return EntryNodeSearchResult()
+
+    PAGE_LENGTH = 50
+    skip = max(0, (page - 1) * PAGE_LENGTH)
+
+    cypher_query = build_cypher_query(query, skip, PAGE_LENGTH)
```

Comment:
It would be better to collect errors here and eventually send them back.

Reply:
I am not sure it really is that useful to do so: https://github.com/openfoodfacts/taxonomy-editor/pull/438/files#r1525369981